src/main/java/org/clafer/ast/compiler/AstCompiler.java

```diff
@@ -1702,6 +1704,24 @@ public IrExpr visit(AstTransitiveClosure ast, Void a) {
             // Bug.
             throw new AstException("Should not have passed type checking.");
         }
+
+        @Override
+        public IrExpr visit(AstConnected ast, Void a) {
+            IrExpr nodes = compile(ast.getNodes());
+            IrExpr relation = compile(ast.getRelation());
+            if (relation instanceof IrIntArrayExpr) {
+                Type type = getType(ast.getRelation());
+                AstClafer returnType = type.getCommonSupertype().get(1);
+                //TODO fix
+                return connected((IrSetVar) nodes, (IrSetArrayExpr) relation, ast.isDirected());
```

Comment:
Note that relation is of type IrIntArrayExpr so you cannot cast it to IrSetArrayExpr. The correct code is below:

```java
return connected((IrSetVar) nodes, asRelation(relation, getCommonSupertype(ast.getRelation())), ast.isDirected());
```